api/src/main/java/io/minio/CopySource.java

```diff
@@ -0,0 +1,139 @@
+  private String objectName;
+  private Integer offset;
+  private Integer length;
+  private ServerSideEncryption sseSource;
+  private Map<String, String> headerMap = new HashMap<>();
+
+  /**
+   * Create new CopySource for given bucket and object.
+   */
+  public CopySource(String bucketName, String objectName ) {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+  }
+  /**
+   * Create new CopySource for given bucket, object, offset and length.
+   */
+  public CopySource(String bucketName, String objectName, Integer offset, Integer length)
+          throws InvalidArgumentException {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    if ((offset != null && offset < 0) || (length != null && length < 0)) {
+      throw new InvalidArgumentException("Offset/Length can not be negative");
+    }
+    this.offset = offset;
+    this.length = length;
+  }
+
+  /**
+   * Create new CopySource for given bucket, object, offset, length and headerMap .
+   */
+  public CopySource(String bucketName, String objectName, Integer offset, Integer length,
+                    Map<String, String> headerMap) throws InvalidArgumentException {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    if ((offset != null && offset < 0) || (length != null && length < 0)) {
+      throw new InvalidArgumentException("Offset/Length can not be negative");
+    }
+    this.offset = offset;
+    this.length = length;
+    this.headerMap = headerMap;
+  }
+
+
+  /**
+   * Create new CopySource for given bucket, object, offset, length, headerMap  and server side encryption .
+   */
+  public CopySource(String bucketName, String objectName, Integer offset, Integer length,
+                    ServerSideEncryption sseSource, Map<String, String> headerMap)
+  throws InvalidArgumentException {
+    this.bucketName = bucketName;
+    this.objectName = objectName;
+    if ((offset != null && offset < 0) || (length != null && length < 0)) {
+      throw new InvalidArgumentException("Offset/Length can not be negative");
+    }
+    this.offset = offset;
+    this.length = length;
+    this.headerMap = headerMap;
+    this.sseSource = sseSource;
+  }
+
+  /**
+   * Returns bucket name.
+   */
+  public String bucketName() {
+    return bucketName;
+  }
+
+  /**
+   * Returns object name.
+   */
+  public String objectName() {
+    return objectName;
+  }
+
+  /**
+   * Returns offset.
+   */
+  public Integer offset() {
+    return offset;
+  }
+
+  /**
+   * Returns length.
+   */
+  public Integer length() {
+    return length;
+  }
+  /**
+   * Returns headerMap.
+   */
+
+  public Map<String, String> getHeaders() {
+    if ( headerMap ==   null ) {
+      headerMap = new HashMap<String,String>();
+    }
+    if ( sseSource != null) {
+      sseSource.marshal(headerMap);
+    }
+    return  headerMap;
+  }
+
+  /**
+   * Sets the range for copy .
+   */
+  public Map<String,String> setCopySourceRange() {
```

Comment:
Few correction in this class.

1. offset and length needs to be `Long`sseSource.marshal(headerMap);
2. all constructors need to call `public CopySource(String bucketName, String objectName, Integer offset, Integer length, ServerSideEncryption sseSource, Map<String, String> headerMap)` to avoid repeated validations.
3. rename `getHeaders()` to `headers()`
4. `sseSource.marshal(headerMap);` needs to be set in constructor `public CopySource(String bucketName, String objectName, Integer offset, Integer length, ServerSideEncryption sseSource, Map<String, String> headerMap)` to make `headers()` to be simple and straight forward.
5. `x-amz-copy-range-source-range` needs to be set in constructor `public CopySource(String bucketName, String objectName, Integer offset, Integer length, ServerSideEncryption sseSource, Map<String, String> headerMap)` and `setCopySourceRange()` is unnecessary.

Reply:
1. Done
2. Done
3. done
4. Done
5.private method `getByteRange` introduced to calculate the value for `x-amz-copy-range-source-range` so it is not in constructor.